File: apps/web/lib/config/aws.ts
type AwsConfig = {
  region: string
  endpoint?: string
  credentials?: { accessKeyId: string; secretAccessKey: string }
}

function isFlociEndpoint(ep: string): boolean {
  return ep.includes("localhost:4566") || ep.includes("127.0.0.1:4566") || ep.includes("host.docker.internal:4566")
}

function envOr(key: string, fallback: string): string {
  const v = process.env[key]
  return v !== undefined && v !== "" ? v : fallback
}

function getAwsConfig(): AwsConfig {
  let region = envOr("AWS_REGION", "")
  if (!region) region = "ap-south-1"
  const endpoint = process.env.AWS_ENDPOINT_URL || undefined
  let credentials: AwsConfig["credentials"] | undefined
  if (!process.env.AWS_ACCESS_KEY_ID && endpoint && isFlociEndpoint(endpoint)) {
    credentials = { accessKeyId: "test", secretAccessKey: "test" }
  }
  return { region, endpoint, credentials }
}

function isMissingSecret(err: unknown): boolean {
  const msg = String((err as Error)?.message ?? err ?? "")
  return msg.includes("ResourceNotFoundException") || msg.includes("not found")
}

function isMissingParam(err: unknown): boolean {
  const msg = String((err as Error)?.message ?? err ?? "")
  return msg.includes("ParameterNotFound") || msg.includes("not found")
}

const KNOWN_SSM_KEYS = new Set([
  "DATABASE_URL",
  "DATABASE_URL_REPLICA",
  "REDIS_URL",
  "REDIS_PASSWORD",
  "NEXTAUTH_SECRET",
  "NEXTAUTH_URL",
  "GOOGLE_ID",
  "GOOGLE_SECRET",
  "GITHUB_ID",
  "GITHUB_SECRET",
  "TURNSTILE_SECRET_KEY",
  "NEXT_PUBLIC_TURNSTILE_SITE_KEY",
  "NEXT_PUBLIC_PADDLE_CLIENT_TOKEN",
  "FILE_EXTRACTOR_API_URL",
  "AI_SERVICE_URL",
  "AI_SERVICE_API_KEY",
  "CHAT_SERVICE_URL",
  "PAYMENT_GATEWAY_URL",
  "INTERNAL_API_KEY",
  "RABBITMQ_URL",
  "PROMETHEUS_WEB_SCRAPE_TOKEN",
  "DB_POOL_MAX",
  "LOG_LEVEL",
  "NODE_ENV",
  "OTEL_EXPORTER_OTLP_ENDPOINT",
  "OTEL_SERVICE_NAME",
  "AWS_REGION",
])

function applySecretJson(data: Record<string, string>, mapping: Record<string, string>, cfg: Record<string, string>): void {
  for (const [rawKey, val] of Object.entries(data)) {
    if (!val) continue
    const upper = rawKey.toUpperCase()
    const targetFromMapping = mapping[upper]
    if (targetFromMapping) {
      if (!cfg[targetFromMapping]) cfg[targetFromMapping] = String(val)
      continue
    }
    if (KNOWN_SSM_KEYS.has(upper) && !cfg[upper]) {
      cfg[upper] = String(val)
    }
  }
}

async function loadSecretsManager(cfg: Record<string, string>): Promise<void> {
  const { SecretsManagerClient, GetSecretValueCommand } = await import("@aws-sdk/client-secrets-manager")
  const awsCfg = getAwsConfig()
  const client = new SecretsManagerClient({
    region: awsCfg.region,
    ...(awsCfg.endpoint ? { endpoint: awsCfg.endpoint } : {}),
    ...(awsCfg.credentials ? { credentials: awsCfg.credentials } : {}),
  })

  const secrets: Array<{ secretId: string; mapping: Record<string, string> }> = [
    {
      secretId: envOr("PG_URLS_SECRET_NAME", "detectai/pg/urls"),
      mapping: { DATABASE_URL: "DATABASE_URL", DATABASE_URL_REPLICA: "DATABASE_URL_REPLICA" },
    },
    {
      secretId: envOr("REDIS_USERS_URLS_SECRET_NAME", envOr("REDIS_URLS_SECRET_NAME", "detectai/redis/users/urls")),
      mapping: { REDIS_URL: "REDIS_URL", REDIS_PASSWORD: "REDIS_PASSWORD" },
    },
    {
      secretId: envOr("MQ_URLS_SECRET_NAME", envOr("MQ_SECRETS_NAME", "detectai/mq/urls")),
      mapping: { RABBITMQ_URL: "RABBITMQ_URL", RABBITMQ_ENDPOINT: "RABBITMQ_URL" },
    },
    {
      secretId: envOr("WEB_SECRETS_NAME", envOr("WEB_SECRETS_ARN", "detectai/web/secrets")),
      mapping: {
        NEXTAUTH_SECRET: "NEXTAUTH_SECRET",
        GOOGLE_ID: "GOOGLE_ID",
        GOOGLE_SECRET: "GOOGLE_SECRET",
        GITHUB_ID: "GITHUB_ID",
        GITHUB_SECRET: "GITHUB_SECRET",
        TURNSTILE_SECRET_KEY: "TURNSTILE_SECRET_KEY",
        AI_SERVICE_API_KEY: "AI_SERVICE_API_KEY",
        INTERNAL_API_KEY: "INTERNAL_API_KEY",
        PROMETHEUS_WEB_SCRAPE_TOKEN: "PROMETHEUS_WEB_SCRAPE_TOKEN",
        NEXT_PUBLIC_TURNSTILE_SITE_KEY: "NEXT_PUBLIC_TURNSTILE_SITE_KEY",
        NEXT_PUBLIC_PADDLE_CLIENT_TOKEN: "NEXT_PUBLIC_PADDLE_CLIENT_TOKEN",
      },
    },
  ]

  for (const { secretId, mapping } of secrets) {
    if (!secretId) continue
    try {
      const out = await client.send(new GetSecretValueCommand({ SecretId: secretId }))
      if (!out.SecretString) continue
      const raw = out.SecretString.trim()
      if (!raw) continue
      try {
        const data = JSON.parse(raw) as Record<string, string>
        applySecretJson(data, mapping, cfg)
      } catch {
        const lower = raw.toLowerCase()
        if (lower.startsWith("amqp") || lower.startsWith("postgres") || lower.startsWith("redis")) {
          const firstKey = Object.keys(mapping)[0]
          if (firstKey && !cfg[mapping[firstKey]]) cfg[mapping[firstKey]] = raw
        }
      }
    } catch (err) {
      const hasNeeded = Object.values(mapping).some((k) => cfg[k])
      if (hasNeeded && isMissingSecret(err)) {
        console.warn(`[config] secret not found but cfg already has value, continuing`, { secret: secretId })
        continue
      }
      if (!isMissingSecret(err)) {
        const msg = String((err as Error)?.message ?? err)
        const isCredsError = msg.includes("security token") || msg.includes("credentials") || msg.includes("ECONNREFUSED")
        if (isCredsError && hasNeeded) {
          console.warn(`[config] AWS unreachable but cfg has values, continuing`, { secret: secretId })
          continue
        }
        throw new Error(`get secret ${secretId}: ${msg}`)
      }
      if (process.env.ENV_TYPE === "prod") {
        const isPrimaryWebSecret = secretId === envOr("WEB_SECRETS_NAME", "detectai/web/secrets")
        if (isPrimaryWebSecret || hasNeeded) {
          console.warn(`[config] prod: web secret not found, continuing without it`, { secret: secretId })
        } else {
          throw new Error(`get secret ${secretId}: ${String((err as Error)?.message ?? err)}`)
        }
      }
    }
  }
}

async function loadSSMParameters(cfg: Record<string, string>): Promise<void> {
  let prefix = envOr("SSM_PREFIX", envOr("SSM_PARAM_PREFIX", ""))
  if (!prefix) {
    prefix = "/detectai/web/"
    if (process.env.SSM_ENABLED === "0" || process.env.SSM_ENABLED === "false") return
  }
  if (!prefix.endsWith("/")) prefix += "/"

  const { SSMClient, GetParametersByPathCommand } = await import("@aws-sdk/client-ssm")
  const awsCfg = getAwsConfig()
  const client = new SSMClient({
    region: awsCfg.region,
    ...(awsCfg.endpoint ? { endpoint: awsCfg.endpoint } : {}),
    ...(awsCfg.credentials ? { credentials: awsCfg.credentials } : {}),
  })

  let nextToken: string | undefined
  let found = false
  do {
    try {
      const out = await client.send(
        new GetParametersByPathCommand({
          Path: prefix,
          Recursive: true,
          WithDecryption: true,
          NextToken: nextToken,
        }),
      )
      nextToken = out.NextToken
      for (const p of out.Parameters ?? []) {
        if (!p.Name || !p.Value) continue
        found = true
        const key = p.Name.slice(prefix.length).toUpperCase().replaceAll("-", "_")
        const val = p.Value
        if (process.env[key] !== undefined && process.env[key] !== "") continue
        if (cfg[key] !== undefined && cfg[key] !== "") continue
        if (key === "POOL_MAX" && !cfg.DB_POOL_MAX) {
          cfg.DB_POOL_MAX = val
          continue
        }
        if (KNOWN_SSM_KEYS.has(key) && !cfg[key]) cfg[key] = val
        else if (!KNOWN_SSM_KEYS.has(key) && !cfg[key]) cfg[key] = val
      }
    } catch (err) {
      if (isMissingParam(err)) return
      throw new Error(`ssm get parameters by path ${prefix}: ${String((err as Error)?.message ?? err)}`)
    }
  } while (nextToken)

  if (!found) return
}

export async function loadFromAWS(cfg: Record<string, string>): Promise<void> {
  await loadSecretsManager(cfg)
  await loadSSMParameters(cfg)
}

File: services/chats/internal/config/aws.go
package config

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"strconv"
	"strings"
	"time"

	"github.com/aws/aws-sdk-go-v2/aws"
	awsconfig "github.com/aws/aws-sdk-go-v2/config"
	"github.com/aws/aws-sdk-go-v2/service/secretsmanager"
	"github.com/aws/aws-sdk-go-v2/service/ssm"
)

func loadFromAWS(ctx context.Context, cfg *Config) error {
	region := cfg.AWSRegion
	if region == "" {
		region = os.Getenv("AWS_REGION")
		if region == "" {
			region = "ap-south-1"
		}
	}
	opts := []func(*awsconfig.LoadOptions) error{awsconfig.WithRegion(region)}
	if ep := os.Getenv("AWS_ENDPOINT_URL"); ep != "" {
		opts = append(opts, awsconfig.WithBaseEndpoint(ep))
	}
	if _, ok := os.LookupEnv("AWS_ACCESS_KEY_ID"); !ok && isFlociEndpoint(os.Getenv("AWS_ENDPOINT_URL")) {
		opts = append(opts, awsconfig.WithCredentialsProvider(aws.CredentialsProviderFunc(func(ctx context.Context) (aws.Credentials, error) {
			return aws.Credentials{AccessKeyID: "test", SecretAccessKey: "test"}, nil
		})))
	}
	awsCfg, err := awsconfig.LoadDefaultConfig(ctx, opts...)
	if err != nil {
		return fmt.Errorf("load AWS config: %w", err)
	}
	if err := loadSecretsManager(ctx, awsCfg, cfg); err != nil {
		return err
	}
	if err := loadSSMParameters(ctx, awsCfg, cfg); err != nil {
		return err
	}
	return nil
}

func loadSecretsManager(ctx context.Context, awsCfg aws.Config, cfg *Config) error {
	sm := secretsmanager.NewFromConfig(awsCfg)

	// DocumentDB URLs (mongo) — Terraform: detectai/docdb/urls
	docdbSecret := envOr("DOCDB_URLS_SECRET_NAME", envOr("DOCDB_SECRETS_NAME", "detectai/docdb/urls"))
	// Allow legacy override via MONGO_URLS_SECRET_NAME
	if v := os.Getenv("MONGO_URLS_SECRET_NAME"); v != "" {
		docdbSecret = v
	}
	if docdbSecret != "" {
		out, err := sm.GetSecretValue(ctx, &secretsmanager.GetSecretValueInput{
			SecretId: aws.String(docdbSecret),
		})
		if err == nil && out.SecretString != nil {
			secretStr := strings.TrimSpace(*out.SecretString)
			if secretStr != "" {
				var data map[string]string
				if jsonErr := json.Unmarshal([]byte(secretStr), &data); jsonErr == nil {
					applySecretMap(data, cfg)
				} else if strings.HasPrefix(strings.ToLower(secretStr), "mongodb") {
					if cfg.MongoURI == "" {
						cfg.MongoURI = secretStr
					}
				}
			}
		} else if err != nil && !isMissingSecret(err) {
			return fmt.Errorf("get secret %s: %w", docdbSecret, err)
		} else if err != nil && cfg.IsProd() {
			return fmt.Errorf("get secret %s: %w", docdbSecret, err)
		}
	}

	// Redis chat URLs — Terraform: detectai/redis/chat/urls
	redisSecret := envOr("REDIS_URLS_SECRET_NAME", envOr("REDIS_CHAT_URLS_SECRET_NAME", envOr("CHAT_REDIS_URLS_SECRET_NAME", "detectai/redis/chat/urls")))
	if v := os.Getenv("REDIS_SECRETS_NAME"); v != "" {
		redisSecret = v
	}
	if redisSecret != "" && redisSecret != docdbSecret {
		out, err := sm.GetSecretValue(ctx, &secretsmanager.GetSecretValueInput{
			SecretId: aws.String(redisSecret),
		})
		if err == nil && out.SecretString != nil {
			secretStr := strings.TrimSpace(*out.SecretString)
			if secretStr != "" {
				var data map[string]string
				if jsonErr := json.Unmarshal([]byte(secretStr), &data); jsonErr == nil {
					applySecretMap(data, cfg)
				} else if strings.HasPrefix(strings.ToLower(secretStr), "redis") || strings.HasPrefix(strings.ToLower(secretStr), "rediss") {
					if cfg.RedisAddr == "" {
						cfg.RedisAddr = secretStr
					}
				}
			}
		} else if err != nil && !isMissingSecret(err) {
			return fmt.Errorf("get secret %s: %w", redisSecret, err)
		} else if err != nil && cfg.IsProd() {
			return fmt.Errorf("get secret %s: %w", redisSecret, err)
		}
	}

	return nil
}

func applySecretMap(data map[string]string, cfg *Config) {
	for k, v := range data {
		if v == "" {
			continue
		}
		upper := strings.ToUpper(strings.TrimSpace(k))
		switch upper {
		case "MONGO_URI":
			if cfg.MongoURI == "" {
				cfg.MongoURI = v
			}
		case "MONGO_DATABASE":
			if cfg.MongoDatabase == "" {
				cfg.MongoDatabase = v
			}
		case "MONGO_MODE":
			if cfg.MongoMode == "" {
				cfg.MongoMode = v
			}
		case "CHAT_REDIS_ADDR":
			if cfg.RedisAddr == "" {
				cfg.RedisAddr = v
			}
		case "REDIS_URL":
			if cfg.RedisAddr == "" {
				cfg.RedisAddr = v
			}
		case "REDIS_URL_READER":
			// Ignore reader — chats always uses primary
		case "REDIS_PASSWORD":
			if cfg.RedisPassword == "" {
				cfg.RedisPassword = v
			}
		case "REDIS_TLS_ENABLED":
			if cfg.RedisTLSEnabled == false {
				if b, err := strconv.ParseBool(v); err == nil {
					cfg.RedisTLSEnabled = b
				} else if strings.EqualFold(v, "true") || v == "1" {
					cfg.RedisTLSEnabled = true
				}
			}
		case "MONGO_TLS_ENABLED":
			if !cfg.MongoTLSEnabled {
				if b, err := strconv.ParseBool(v); err == nil {
					cfg.MongoTLSEnabled = b
				}
			}
		case "REDIS_ADDR":
			if cfg.RedisAddr == "" {
				cfg.RedisAddr = v
			}
		default:
			// Generic: allow SSM-like keys to populate if they match known fields and are not already set.
			// This keeps forward-compat for Terraform adding new keys.
			switch upper {
			case "REDIS_ADDR":
				if cfg.RedisAddr == "" {
					cfg.RedisAddr = v
				}
			}
		}
	}
}

func isFlociEndpoint(ep string) bool {
	return strings.Contains(ep, "localhost:4566") || strings.Contains(ep, "127.0.0.1:4566") || strings.Contains(ep, "host.docker.internal:4566")
}

func loadSSMParameters(ctx context.Context, awsCfg aws.Config, cfg *Config) error {
	prefix := envOr("SSM_PREFIX", envOr("SSM_PARAM_PREFIX", ""))
	if prefix == "" {
		prefix = "/detectai/chat/"
		if os.Getenv("SSM_ENABLED") == "0" || strings.EqualFold(os.Getenv("SSM_ENABLED"), "false") {
			return nil
		}
	}
	if !strings.HasSuffix(prefix, "/") {
		prefix += "/"
	}
	client := ssm.NewFromConfig(awsCfg)
	paginator := ssm.NewGetParametersByPathPaginator(client, &ssm.GetParametersByPathInput{
		Path:           aws.String(prefix),
		Recursive:      aws.Bool(true),
		WithDecryption: aws.Bool(true),
	})
	found := false
	for paginator.HasMorePages() {
		out, err := paginator.NextPage(ctx)
		if err != nil {
			if isMissingParam(err) {
				return nil
			}
			return fmt.Errorf("ssm get parameters by path %s: %w", prefix, err)
		}
		for _, p := range out.Parameters {
			if p.Name == nil || p.Value == nil {
				continue
			}
			found = true
			key := strings.TrimPrefix(*p.Name, prefix)
			key = strings.ToUpper(strings.ReplaceAll(key, "-", "_"))
			val := *p.Value
			switch key {
			case "MONGO_URI":
				if os.Getenv("MONGO_URI") == "" && cfg.MongoURI == "" {
					cfg.MongoURI = val
				}
			case "MONGO_DATABASE":
				if os.Getenv("MONGO_DATABASE") == "" && cfg.MongoDatabase == "" {
					cfg.MongoDatabase = val
				}
			case "MONGO_MODE":
				if os.Getenv("MONGO_MODE") == "" && cfg.MongoMode == "" {
					cfg.MongoMode = val
				}
			case "MONGO_TLS_ENABLED":
				if os.Getenv("MONGO_TLS_ENABLED") == "" {
					if b, err := strconv.ParseBool(val); err == nil {
						cfg.MongoTLSEnabled = b
					}
				}
			case "MONGO_TLS_CA_FILE":
				if os.Getenv("MONGO_TLS_CA_FILE") == "" && cfg.MongoTLSCAFile == "" {
					cfg.MongoTLSCAFile = val
				}
			case "MONGO_MAX_POOL_SIZE":
				if os.Getenv("MONGO_MAX_POOL_SIZE") == "" && cfg.MongoMaxPoolSize == 0 {
					if u, err := strconv.ParseUint(val, 10, 64); err == nil {
						cfg.MongoMaxPoolSize = u
					}
				}
			case "MONGO_MIN_POOL_SIZE":
				if os.Getenv("MONGO_MIN_POOL_SIZE") == "" && cfg.MongoMinPoolSize == 0 {
					if u, err := strconv.ParseUint(val, 10, 64); err == nil {
						cfg.MongoMinPoolSize = u
					}
				}
			case "MONGO_SERVER_SELECTION_TIMEOUT":
				if os.Getenv("MONGO_SERVER_SELECTION_TIMEOUT") == "" && cfg.MongoServerTimeout == 0 {
					if d, err := time.ParseDuration(val); err == nil {
						cfg.MongoServerTimeout = d
					}
				}
			case "CHAT_REDIS_ADDR":
				if os.Getenv("CHAT_REDIS_ADDR") == "" && os.Getenv("REDIS_URL") == "" && cfg.RedisAddr == "" {
					cfg.RedisAddr = val
				}
			case "REDIS_URL":
				if os.Getenv("REDIS_URL") == "" && os.Getenv("CHAT_REDIS_ADDR") == "" && cfg.RedisAddr == "" {
					cfg.RedisAddr = val
				}
			case "REDIS_PASSWORD":
				if os.Getenv("REDIS_PASSWORD") == "" && cfg.RedisPassword == "" {
					cfg.RedisPassword = val
				}
			case "REDIS_TLS_ENABLED":
				if os.Getenv("REDIS_TLS_ENABLED") == "" {
					if b, err := strconv.ParseBool(val); err == nil {
						cfg.RedisTLSEnabled = b
					}
				}
			case "REDIS_TLS_CA_FILE":
				if os.Getenv("REDIS_TLS_CA_FILE") == "" && cfg.RedisTLSCAFile == "" {
					cfg.RedisTLSCAFile = val
				}
			case "REDIS_POOL_SIZE":
				if os.Getenv("REDIS_POOL_SIZE") == "" && cfg.RedisPoolSize == 0 {
					if i, err := strconv.Atoi(val); err == nil {
						cfg.RedisPoolSize = i
					}
				}
			case "BATCH_SIZE":
				if os.Getenv("BATCH_SIZE") == "" && cfg.BatchSize == 0 {
					if i, err := strconv.Atoi(val); err == nil {
						cfg.BatchSize = i
					}
				}
			case "STREAM_PARTITION_COUNT":
				if os.Getenv("STREAM_PARTITION_COUNT") == "" && cfg.StreamPartitionCount == 0 {
					if i, err := strconv.Atoi(val); err == nil {
						cfg.StreamPartitionCount = i
					}
				}
			case "CACHE_TTL":
				if os.Getenv("CACHE_TTL") == "" && cfg.CacheTTL == 0 {
					if d, err := time.ParseDuration(val); err == nil {
						cfg.CacheTTL = d
					}
				}
			case "GRPC_PORT":
				if os.Getenv("GRPC_PORT") == "" && cfg.GRPCPort == "" {
					cfg.GRPCPort = val
				}
			case "METRICS_PORT":
				if os.Getenv("METRICS_PORT") == "" && cfg.MetricsPort == "" {
					cfg.MetricsPort = val
				}
			case "LOG_LEVEL":
				if os.Getenv("LOG_LEVEL") == "" && cfg.LogLevel == "" {
					cfg.LogLevel = val
				}
			case "OTEL_EXPORTER_OTLP_ENDPOINT":
				if os.Getenv("OTEL_EXPORTER_OTLP_ENDPOINT") == "" && cfg.OtelEndpoint == "" {
					cfg.OtelEndpoint = val
				}
			case "OTEL_SERVICE_NAME":
				if os.Getenv("OTEL_SERVICE_NAME") == "" && cfg.OtelServiceName == "" {
					cfg.OtelServiceName = val
				}
			case "AWS_REGION":
				if os.Getenv("AWS_REGION") == "" && cfg.AWSRegion == "" {
					cfg.AWSRegion = val
				}
			default:
				// Allow any additional SSM key to populate if env not set and cfg empty —
				// keeps parity with workers/document-parser extra=ignore behaviour.
				// No-op for unknown keys.
			}
		}
	}
	if !found {
		return nil
	}
	return nil
}

func isMissingSecret(err error) bool {
	if err == nil {
		return false
	}
	msg := err.Error()
	return strings.Contains(msg, "ResourceNotFoundException") || strings.Contains(msg, "not found")
}

func isMissingParam(err error) bool {
	if err == nil {
		return false
	}
	msg := err.Error()
	return strings.Contains(msg, "ParameterNotFound") || strings.Contains(msg, "not found")
}
